File: sdks/kotlin/src/main/kotlin/to/llmux/kotlin/Direct.kt
@file:JvmName("LlmuxDirectKt")

package to.llmux.kotlin

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.channels.trySendBlocking
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.buffer
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.launch
import to.llmux.LlmuxDirect
import java.nio.file.Path
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Kotlin over [to.llmux.LlmuxDirect] — llmux running **in this JVM** through
 * libllmux's C ABI.
 *
 * This is a thin, idiomatic layer, not a reimplementation. The FFM binding, the
 * memory rules and the handle lifecycle all live in the Java class; what Kotlin
 * adds is `use {}`, named arguments, and a [Flow] for streaming that cancels
 * the underlying Go stream when the collector stops caring.
 *
 * **The sidecar is the recommended default on the JVM.** See `README.md` →
 * "The JVM and Go's signal handlers"; the argument is measured, and it applies
 * to Kotlin exactly as it does to Java, because it is an argument about the
 * process, not the language.
 *
 * ```kotlin
 * Llmux.direct(configJson).use { llmux ->
 *     println(llmux.call("models"))
 *     llmux.chunks("chat", requestJson).collect { print(it) }
 * }
 * ```
 *
 * Requires Java 22+ and `--enable-native-access=ALL-UNNAMED`.
 */
public class LlmuxGateway internal constructor(
    private val delegate: LlmuxDirect,
) : AutoCloseable {

    /** The llmux version the loaded shared library was built from. */
    public val abiVersion: String get() = delegate.abiVersion()

    /** The library this gateway is running out of. */
    public val libraryPath: Path get() = delegate.libraryPath()

    /**
     * One unary call. [method] is `"chat"`, `"embed"` or `"models"`.
     *
     * @throws to.llmux.LlmuxException carrying the library's own message
     */
    public fun call(method: String, requestJson: String? = null): String =
        delegate.call(method, requestJson)

    /**
     * Stream a chat completion, blocking this thread and invoking [onChunk]
     * once per chunk on it. Return `false` from [onChunk] to stop; stopping is
     * a decision, not an error.
     *
     * Prefer [chunks] unless you are deliberately staying off coroutines. Note
     * that the underlying [to.llmux.LlmuxDirect.stream] this delegates to
     * declares a checked `InterruptedException` — Kotlin does not enforce
     * catching it, but it can still reach you here if this thread is
     * interrupted while blocked; see [to.llmux.LlmuxDirect.stream]'s
     * "Cancellation" section (in the Java SDK) for why, and use [chunks]
     * instead if you want cancellation as `CancellationException`.
     *
     * @return the number of chunks delivered
     */
    public fun stream(
        method: String = "chat",
        requestJson: String,
        onChunk: (String) -> Boolean,
    ): Int = delegate.stream(method, requestJson, onChunk)

    /**
     * Aborts every call in flight on this gateway — [call] and [stream] (and
     * every in-progress [chunks] collection) alike — without closing it. A
     * thin delegate to [to.llmux.LlmuxDirect.cancel]; read that method's doc
     * for the full contract, in particular that **it is per-HANDLE, not
     * per-call**: it also aborts a sibling [call] or [stream] running
     * concurrently on this same [LlmuxGateway]. [chunks] calls this itself
     * when its `Flow` is cancelled or abandoned, so most callers never reach
     * for it directly — it exists at this level for the same reason
     * `to.llmux.LlmuxDirect.cancel` does: something else, entirely outside any
     * one `Flow`, decided this gateway should stop.
     */
    public fun cancel(): Unit = delegate.cancel()

    /**
     * The same stream as a cold [Flow] of chunk JSON documents, cancellable
     * the way coroutines expect: cancelling the collecting coroutine, the
     * enclosing scope, or a `withTimeout` around the collection reaches
     * `llmux_cancel`, and what a cancelled collector sees is
     * [kotlinx.coroutines.CancellationException] — never a wrapped
     * `LlmuxException` complaining "context canceled" about a cancellation
     * this Flow caused on its own behalf.
     *
     * Four properties worth knowing:
     *
     * * **Cancellation reaches `llmux_cancel` directly, not just by
     *   abandonment.** [awaitClose] calls [to.llmux.LlmuxDirect.cancel]
     *   itself, rather than relying solely on the next chunk noticing the
     *   channel is closed and returning "stop" from the callback. The
     *   difference matters when the provider is *between* chunks — mid
     *   network read, no callback pending: without an explicit `cancel()`
     *   there and then, `llmux_stream` keeps blocking until the provider
     *   produces another chunk for the callback to notice on, which could be
     *   a long time. `cancel()` unblocks it immediately instead, from another
     *   thread — which its own contract calls safe.
     * * **That is why this flow is also a RENDEZVOUS, not a buffer**, which
     *   still matters independently of the point above: even with `cancel()`
     *   wired in, a *buffered* channel lets the callback keep accepting
     *   chunks the collector already stopped wanting, right up until the
     *   buffer fills. Measured against the fake upstream, on a five-chunk
     *   stream collected with `take(2)`: with `Channel.BUFFERED` the callback
     *   fired **5** times — the producer filled the 64-slot buffer before the
     *   collector ever cancelled, so `take` discarded three chunks that had
     *   already been generated and billed. With `Channel.RENDEZVOUS` it fired
     *   **3**: two collected and one in flight at the handoff. The cost is one
     *   context switch per chunk, which against a model emitting tokens is
     *   free.
     * * **The blocking call runs on [Dispatchers.IO], and this is what makes
     *   cancellation observable at all.** `llmux_stream` blocks its thread for
     *   the life of the stream — it has no suspension point of its own — so if
     *   it ran on the collector's dispatcher, cancelling the collector would
     *   have nothing to interrupt: the thread running the blocking call would
     *   never be freed up to notice. Running it on its own `Dispatchers.IO`
     *   thread means the *producer's* cancellation (in [awaitClose], which
     *   fires as soon as the collector or its scope cancels, independent of
     *   what the IO thread is doing) is what reaches `cancel()`, not something
     *   the blocked thread has to cooperate with on its own.
     * * A cancellation this `Flow` caused on itself — via `awaitClose` — is
     *   swallowed rather than handed to the collector as a `LlmuxException`:
     *   `awaitClose` sets a flag before calling `cancel()`, and the internal
     *   job's catch block checks that flag before deciding whether to close
     *   the channel with the resulting error or cleanly. An *external*,
     *   unrelated call to [cancel] — fact 6 in the C ABI's own docs,
     *   "per-handle, not per-call" — still surfaces as a real
     *   `LlmuxException`, because from this `Flow`'s perspective that is a
     *   genuine disruption, not its own shutdown.
     *
     * Every chunk is one `chat.completion.chunk` object — the same JSON the
     * HTTP API writes after `data: `. Note that one chunk beyond your last
     * collected one may already have been generated: cancellation is prompt,
     * not retroactive.
     */
    public fun chunks(
        method: String = "chat",
        requestJson: String,
    ): Flow<String> = callbackFlow {
        // Set BEFORE cancel() below, not inferred afterward from isActive: a
        // flag written first and read once, on the thread that is actually
        // deciding, is simpler to reason about than a race between two
        // notifications (job cancellation, native call unwinding) landing in
        // an unpredictable order.
        val cancelledByAwaitClose = AtomicBoolean(false)

        // llmux_stream blocks, so it gets its own IO thread; the flow's
        // producer scope stays free to observe cancellation.
        val job = launch(Dispatchers.IO) {
            try {
                delegate.stream(method, requestJson) { chunk ->
                    // Backpressure with a stop signal: a failed send means the
                    // collector is gone, which is exactly when to stop. This
                    // is the FALLBACK path — cancel() below is the primary
                    // one — for the case where a chunk was already in flight
                    // when cancellation was requested.
                    trySendBlocking(chunk).isSuccess
                }
                close()
            } catch (t: Throwable) {
                // llmux_cancel has no concept of Kotlin cancellation: our own
                // call to it below surfaces here as an ordinary
                // LlmuxException ("context canceled"), exactly as an
                // unrelated external cancel() would. Distinguish them by
                // whether WE asked for this: if awaitClose already fired, the
                // structured-concurrency machinery is already delivering
                // CancellationException to the collector on its own, and
                // handing it a second, different exception through the
                // channel would replace that with a confusing wrapped error.
                if (cancelledByAwaitClose.get()) close() else close(t)
            }
        }
        awaitClose {
            cancelledByAwaitClose.set(true)
            // Reach llmux_cancel directly — see the third bullet above — then
            // cancel the coroutine job for ordinary structured-concurrency
            // bookkeeping. Order matters: the flag must be visible before the
            // native call can possibly fail because of it.
            delegate.cancel()
            job.cancel()
        }
    }.buffer(Channel.RENDEZVOUS).flowOn(Dispatchers.IO)

    /** Release the gateway. Idempotent; `use {}` calls it on every path out. */
    override fun close(): Unit = delegate.close()
}

/** Entry points for both llmux paths. */
public object Llmux {

    /**
     * Open an in-process gateway.
     *
     * @param configJson an llmux configuration document, or null for built-in
     *                   defaults plus the environment
     * @param library    an explicit libllmux; defaults to
     *                   [LlmuxDirect.findLibrary]'s search
     */
    public fun direct(configJson: String? = null, library: Path? = null): LlmuxGateway =
        LlmuxGateway(
            if (library == null) LlmuxDirect.open(configJson)
            else LlmuxDirect.open(library, configJson),
        )

    /** Where the direct path would load its library from, without loading it. */
    public fun findLibrary(): Path = LlmuxDirect.findLibrary()
}

File: sdks/swift/Sources/LLMux/Direct.swift
// Direct mode: llmux running inside this process, over the C ABI.
//
// Swift's C interop is one of its genuine strengths, and this file uses the
// plainest form of it: dlopen/dlsym plus @convention(c) function types. No
// module map, no bridging header, no unsafeFlags in Package.swift — which
// matters, because a target with unsafeFlags cannot be depended on by another
// package.
//
// SPDX-License-Identifier: MIT OR Apache-2.0

#if canImport(Darwin)
import Darwin
#else
import Glibc
#endif
import Foundation

// MARK: - Errors

/// Everything that can go wrong on the direct path.
public enum LLMuxError: Error, CustomStringConvertible {
    /// No `libllmux` could be found. Carries the paths that were tried.
    case libraryNotFound([String])
    /// `dlopen` failed. Carries the platform's message.
    case load(String)
    /// A symbol was missing — usually a library that is not llmux, or one from
    /// a different major version.
    case missingSymbol(String)
    /// The loaded library reports a different version than expected.
    case versionMismatch(loaded: String, expected: String)
    /// llmux itself failed. The string is the library's own message, which is
    /// plain UTF-8 text and deliberately **not** JSON — do not parse it.
    case llmux(String)

    public var description: String {
        switch self {
        case .libraryNotFound(let tried):
            return """
                libllmux not found. Set LLMUX_LIBRARY to its path, or build it with \
                `scripts/build-ffi.sh`. Tried: \(tried.joined(separator: ", "))
                """
        case .load(let msg):
            return "loading libllmux: \(msg)"
        case .missingSymbol(let name):
            return "libllmux is missing the symbol \(name) — is this really libllmux?"
        case .versionMismatch(let loaded, let expected):
            return """
                libllmux reports version \(loaded), expected \(expected) — a stale library is \
                earlier on your load path
                """
        case .llmux(let msg):
            return msg
        }
    }
}

// MARK: - The ABI

typealias AbiVersionFn = @convention(c) () -> UnsafePointer<CChar>?
typealias NewFn = @convention(c) (
    UnsafePointer<CChar>?, UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UInt64
typealias CloseFn = @convention(c) (UInt64) -> Void
typealias CancelFn = @convention(c) (UInt64) -> Void
typealias CallFn = @convention(c) (
    UInt64, UnsafePointer<CChar>?, UnsafePointer<CChar>?,
    UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>?
typealias FreeFn = @convention(c) (UnsafeMutablePointer<CChar>?) -> Void
typealias ChunkCb = @convention(c) (UnsafePointer<CChar>?, UnsafeMutableRawPointer?) -> Int32
typealias StreamFn = @convention(c) (
    UInt64, UnsafePointer<CChar>?, UnsafePointer<CChar>?, ChunkCb?, UnsafeMutableRawPointer?,
    UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32

/// The loaded library and its seven resolved symbols.
///
/// # Why this is never unloaded
///
/// There is no `dlclose` anywhere in this file, and that is deliberate.
/// `libllmux` is a Go `c-shared` object: loading it starts the Go runtime and
/// its threads, and Go has no way to shut that down, so unloading would unmap
/// code those threads are still executing.
///
/// The Rust binding for this same library was written the other way first —
/// unloading with each handle — and a 200-cycle open/close loop hung and had to
/// be killed. `Library.shared(at:)` caches one instance per path for the life
/// of the process.
final class Library: @unchecked Sendable {
    let abiVersion: AbiVersionFn
    let new: NewFn
    let close: CloseFn
    let cancel: CancelFn
    let call: CallFn
    let free: FreeFn
    let stream: StreamFn

    private static let lock = NSLock()
    private static var loaded: [String: Library] = [:]

    /// The process-wide `Library` for `path`, loading it on first use.
    static func shared(at path: String) throws -> Library {
        lock.lock()
        defer { lock.unlock() }
        if let existing = loaded[path] { return existing }
        let lib = try Library(path: path)
        loaded[path] = lib
        return lib
    }

    private init(path: String) throws {
        guard let handle = dlopen(path, RTLD_NOW | RTLD_LOCAL) else {
            let reason = dlerror().map { String(cString: $0) } ?? "unknown error"
            throw LLMuxError.load("\(path): \(reason)")
        }
        // Deliberately not stored: nothing may ever dlclose it. See above.
        func sym<T>(_ name: String, _ type: T.Type) throws -> T {
            guard let p = dlsym(handle, name) else { throw LLMuxError.missingSymbol(name) }
            return unsafeBitCast(p, to: type)
        }
        abiVersion = try sym("llmux_abi_version", AbiVersionFn.self)
        new = try sym("llmux_new", NewFn.self)
        close = try sym("llmux_close", CloseFn.self)
        cancel = try sym("llmux_cancel", CancelFn.self)
        call = try sym("llmux_call", CallFn.self)
        free = try sym("llmux_free", FreeFn.self)
        stream = try sym("llmux_stream", StreamFn.self)
    }

    var version: String {
        // llmux_abi_version returns a STATIC string owned by the library. It
        // must not be freed — the one exception to the ownership rule, and
        // freeing it would be heap corruption that only shows up later.
        guard let p = abiVersion() else { return "" }
        return String(cString: p)
    }

    /// Takes ownership of a `char*` error message, copies it, and **frees it**.
    /// Letting the message reach a Swift `Error` without this is the classic
    /// C-ABI binding leak: error strings are malloc'd exactly like results.
    func takeError(_ err: UnsafeMutablePointer<CChar>?) -> LLMuxError {
        guard let err else { return .llmux("llmux failed without a message") }
        let msg = String(cString: err)
        free(err)
        return .llmux(msg)
    }

    /// Takes ownership of a `char*` result, copies it, and frees it.
    func takeString(_ p: UnsafeMutablePointer<CChar>) -> String {
        let s = String(cString: p)
        free(p)
        return s
    }
}

// MARK: - Locating the library

public enum LibraryLocator {
    /// The platform's file name for the shared library.
    ///
    /// Windows appears here for completeness only: **no `llmux.dll` has ever
    /// been built.** Swift on Windows cannot use direct mode today.
    public static var fileName: String {
        #if os(Windows)
        return "llmux.dll"
        #elseif os(macOS)
        return "libllmux.dylib"
        #else
        return "libllmux.so"
        #endif
    }

    /// Finds `libllmux`, in this order:
    ///
    /// 1. `$LLMUX_LIBRARY`, if set — an explicit path always wins.
    /// 2. `dist/ffi/<goos>_<goarch>/`, walking up from the current directory,
    ///    which is where `scripts/build-ffi.sh` puts it in a checkout.
    /// 3. The bare file name, handed to the platform loader.
    public static func find() throws -> String {
        var tried: [String] = []

        if let explicit = ProcessInfo.processInfo.environment["LLMUX_LIBRARY"], !explicit.isEmpty {
            if FileManager.default.fileExists(atPath: explicit) { return explicit }
            tried.append(explicit)
        }

        #if os(macOS)
        let goos = "darwin"
        #elseif os(Windows)
        let goos = "windows"
        #else
        let goos = "linux"
        #endif
        #if arch(arm64)
        let goarch = "arm64"
        #else
        let goarch = "amd64"
        #endif
        let relative = "dist/ffi/\(goos)_\(goarch)/\(fileName)"

        var dir = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
        while true {
            let candidate = dir.appendingPathComponent(relative).path
            if FileManager.default.fileExists(atPath: candidate) { return candidate }
            tried.append(candidate)
            let parent = dir.deletingLastPathComponent()
            if parent.path == dir.path { break }
            dir = parent
        }

        // Let the loader try the bare name; if it resolves, use it.
        if let h = dlopen(fileName, RTLD_NOW | RTLD_LOCAL) {
            _ = h  // never dlclose: see Library
            return fileName
        }
        tried.append(fileName)
        throw LLMuxError.libraryNotFound(tried)
    }
}

// MARK: - Gateway

/// An llmux gateway running in this process.
///
/// **Closing is `deinit`.** The handle is released when the last reference goes
/// away — on the happy path, on a `throw`, and on an early `return`. Swift's
/// ARC is the RAII here; there is no `close()` to forget, and no `defer` needed
/// at the call site.
///
/// The ABI documents a handle as safe to use from several threads at once, so
/// this type is `Sendable`.
public final class Gateway: @unchecked Sendable {
    private let lib: Library
    private let handle: UInt64

    /// Loads the library (see ``LibraryLocator/find()``) and creates a gateway.
    ///
    /// `configJSON` is an llmux configuration document — the same JSON
    /// `llmux serve` reads from `llmux.json`. `nil` means built-in defaults
    /// plus the environment.
    ///
    /// Construction is **inert**: no goroutines start and, unless the config
    /// names a Postgres DSN, no sockets open.
    public convenience init(configJSON: String? = nil) throws {
        try self.init(libraryPath: try LibraryLocator.find(), configJSON: configJSON)
    }

    /// As above, against a library at a known path.
    public init(libraryPath: String, configJSON: String? = nil) throws {
        let lib = try Library.shared(at: libraryPath)
        self.lib = lib

        var err: UnsafeMutablePointer<CChar>?
        let h: UInt64 = withOptionalCString(configJSON) { cfg in
            lib.new(cfg, &err)
        }
        guard h != 0 else { throw lib.takeError(err) }
        self.handle = h
    }

    /// As above, refusing to continue if the loaded library is not the expected
    /// version. Worth doing at startup: a shared library resolves off a load
    /// path you may not control.
    public convenience init(expectedVersion: String, configJSON: String? = nil) throws {
        let path = try LibraryLocator.find()
        let lib = try Library.shared(at: path)
        let loaded = lib.version
        guard loaded == expectedVersion else {
            throw LLMuxError.versionMismatch(loaded: loaded, expected: expectedVersion)
        }
        try self.init(libraryPath: path, configJSON: configJSON)
    }

    deinit {
        // llmux_close is idempotent and aborts any stream still running on the
        // handle, so this is safe however we got here.
        lib.close(handle)
    }

    /// The version the loaded library was built from, e.g. `"0.1.2"`.
    public var abiVersion: String { lib.version }

    /// Aborts every call in flight on this gateway, without closing it: the
    /// handle stays open and the next call starts on a fresh context.
    ///
    /// `call`, `streamSync`, and therefore `chunks` all block the calling
    /// thread until they finish — this is the only way to abandon one from
    /// outside. Call it from another thread while a call is blocked and the
    /// blocked call returns `LLMuxError.llmux("context canceled")`; chunks
    /// already delivered stay delivered, and tokens already served are still
    /// metered.
    ///
    /// It is also safe to call from **inside a chunk callback** — unlike
    /// `llmux_close`, which must never be called from a callback because it
    /// waits for the very call that is running it, `llmux_cancel` does not
    /// wait. This is the only cancellation path available to a host whose
    /// callback is its only thread (Node, PHP), and it is exactly what
    /// ``chunks(_:)`` relies on being safe: cancelling one `AsyncThrowingStream`
    /// can race a chunk arriving on the background queue that is driving it.
    ///
    /// A no-op when nothing is running on this handle, and a no-op the second
    /// time, so a defensive `cancel()` in a cleanup path costs nothing.
    ///
    /// **This is per-HANDLE, not per-call.** It aborts every call in flight on
    /// this `Gateway`, including one your code did not mean to touch. There is
    /// no finer-grained cancel in the ABI — if you need to cancel one stream
    /// without disturbing another that happens to share a handle, the fix is
    /// one `Gateway` per cancellation scope. ``chunks(_:)`` calls this from
    /// `onTermination`, so cancelling or abandoning one streamed
    /// `AsyncSequence` aborts every other call in flight on the same gateway.
    public func cancel() {
        lib.cancel(handle)
    }

    /// One unary call. `method` is `"chat"`, `"embed"` or `"models"`; the
    /// request and the result are the same JSON the HTTP API uses.
    ///
    /// A `"chat"` request with `"stream": true` is **refused** rather than
    /// quietly served as one blob — use ``chunks(_:)``.
    public func call(_ method: String, _ requestJSON: String? = nil) throws -> String {
        var err: UnsafeMutablePointer<CChar>?
        let out: UnsafeMutablePointer<CChar>? = method.withCString { m in
            withOptionalCString(requestJSON) { req in
                lib.call(handle, m, req, &err)
            }
        }
        guard let out else { throw lib.takeError(err) }
        return lib.takeString(out)
    }

    /// `POST /v1/chat/completions`, non-streaming.
    public func chat(_ requestJSON: String) throws -> String {
        try call("chat", requestJSON)
    }

    /// `POST /v1/embeddings`.
    public func embed(_ requestJSON: String) throws -> String {
        try call("embed", requestJSON)
    }

    /// `GET /v1/models`, answered from memory with no upstream call.
    public func models() throws -> String {
        try call("models", nil)
    }

    // MARK: Streaming

    /// Streams a chat completion, calling `onChunk` once per chunk **on this
    /// thread**, and blocking until the stream ends.
    ///
    /// Return `true` from `onChunk` to keep streaming and `false` to stop.
    /// Stopping early is not an error — you returned `false`, so you already
    /// know — and this returns normally. Tokens already served are metered
    /// either way.
    ///
    /// The chunk string is owned by the library and valid only for the duration
    /// of the callback; this method copies it into a Swift `String` before
    /// handing it over, so the copy is yours.
    ///
    /// Prefer ``chunks(_:)`` unless you specifically want the blocking form.
    public func streamSync(_ requestJSON: String, onChunk: (String) -> Bool) throws {
        // A @convention(c) function cannot capture, so the closure travels
        // through user_data as an unmanaged pointer to a box.
        final class Box {
            let body: (String) -> Bool
            init(_ body: @escaping (String) -> Bool) { self.body = body }
        }
        // withoutActuallyEscaping: the closure does not outlive this call —
        // llmux_stream is documented to invoke it synchronously and to return
        // only once the stream has ended.
        try withoutActuallyEscaping(onChunk) { escaping in
            let box = Box(escaping)
            let user = Unmanaged.passUnretained(box).toOpaque()

            let trampoline: ChunkCb = { chunkPtr, userData in
                guard let userData else { return 1 }
                let box = Unmanaged<Box>.fromOpaque(userData).takeUnretainedValue()
                let text = chunkPtr.map { String(cString: $0) } ?? ""
                return box.body(text) ? 0 : 1
            }

            var err: UnsafeMutablePointer<CChar>?
            let rc: Int32 = requestJSON.withCString { req in
                "chat".withCString { m in
                    lib.stream(handle, m, req, trampoline, user, &err)
                }
            }
            // Keep `box` alive across the call; passUnretained alone does not.
            withExtendedLifetime(box) {}
            if rc != 0 { throw lib.takeError(err) }
        }
    }

    /// Streams a chat completion as an `AsyncSequence` of chunk JSON documents.
    ///
    /// ```swift
    /// for try await chunk in gw.chunks(requestJSON) {
    ///     print(chunk)
    /// }
    /// ```
    ///
    /// Each element is one `chat.completion.chunk` object — byte for byte what
    /// the HTTP API writes after `data: `, without the prefix and without the
    /// terminal `[DONE]` frame.
    ///
    /// # How this works, and what it does not give you
    ///
    /// `llmux_stream` is a **blocking call with a callback**; an `AsyncSequence`
    /// is a pull API. Bridging them needs somewhere for the blocking call to
    /// live, so this dispatches it to a background queue — *not* to the
    /// cooperative thread pool, because blocking one of those threads for the
    /// length of a model's answer is how a Swift concurrency program deadlocks.
    ///
    /// Chunks are yielded **as they arrive**, so time-to-first-token is real.
    /// But `AsyncThrowingStream.yield` never suspends — that is the whole
    /// difference between "buffered" and "backpressured" — so `.unbounded`
    /// buffering means a consumer slower than the model queues chunks in
    /// memory rather than pausing the producer. For a chat completion that is
    /// bounded by the answer length, and this SDK keeps `.unbounded`
    /// deliberately rather than switching to `.bufferingNewest`, which would
    /// silently drop earlier chunks — the wrong trade for text, where every
    /// chunk is content, not a replaceable sample. (The Rust binding for the
    /// same ABI uses a rendezvous channel and gets real backpressure — a
    /// difference in what the two languages' stream types offer, not in the
    /// ABI. The sidecar's stream, below, gets it from `URLSession.bytes(for:)`
    /// driving the socket read.)
    ///
    /// None of that touches the bug this method exists to not have: a
    /// consumer that stops reading must stop the *upstream provider*, not just
    /// stop queuing chunks nobody will read. `AsyncThrowingStream` calls
    /// `onTermination` with `.cancelled` exactly when that happens — the
    /// consuming `Task` is cancelled, or the iterator is simply abandoned by
    /// breaking out of the loop — and this method's `onTermination` reaches
    /// for ``cancel()`` there, not just a flag. A flag only helps once the
    /// *next* chunk arrives and the callback runs again; if the upstream has
    /// gone quiet between chunks, `streamSync` is blocked deep in a network
    /// read on the background queue with nothing scheduled to check a flag.
    /// `llmux_cancel` is the one thing that reaches into that blocked call
    /// from another thread and makes it return now, per fact 1 in the FFI
    /// contract: a cancelled `llmux_stream` returns promptly with
    /// `context canceled`, whatever it was doing.
    ///
    /// That C-ABI string is not what reaches the consumer, though: catching
    /// `LLMuxError.llmux("context canceled")` here and re-throwing
    /// `CancellationError()` is deliberate. `context canceled` is Go's
    /// `context.Canceled.Error()`, produced only when something cancels the
    /// call's context — llmux's own liveness timeouts return their own,
    /// differently worded errors (`stream_idle_timeout_seconds` and
    /// `stream_first_byte_timeout_seconds` in the configuration document), so
    /// this text is never ambiguous with "the model actually failed." A Swift
    /// consumer catching a cancelled `for try await` loop should see the
    /// idiomatic `CancellationError`, not a foreign-language error string it
    /// has no reason to pattern-match on.
    public func chunks(_ requestJSON: String) -> AsyncThrowingStream<String, Error> {
        AsyncThrowingStream(String.self, bufferingPolicy: .unbounded) { continuation in
            continuation.onTermination = { [self] reason in
                switch reason {
                case .cancelled:
                    // The consumer walked away — Task cancellation or a plain
                    // `break` — while `streamSync` below may still be blocked
                    // on the background queue. See ``cancel()`` for what this
                    // does and does not reach: it is per-HANDLE, so it also
                    // aborts any other call sharing this `Gateway`.
                    cancel()
                case .finished:
                    // The producer already returned, through completion or its
                    // own error. Nothing is left in flight to cancel, and
                    // calling `cancel()` anyway would be a no-op — see fact 4 —
                    // but there is no reason to reach for the ABI at all here.
                    break
                @unknown default:
                    cancel()
                }
            }

            DispatchQueue.global(qos: .userInitiated).async { [self] in
                do {
                    try streamSync(requestJSON) { chunk in
                        continuation.yield(chunk)
                        return true
                    }
                    continuation.finish()
                } catch LLMuxError.llmux("context canceled") {
                    continuation.finish(throwing: CancellationError())
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }
}

/// Runs `body` with a C string for `s`, or with `nil` when `s` is `nil`.
///
/// `String.withCString` has no optional form, and the naive workaround —
/// `s?.withCString { … } ?? body(nil)` — does not type-check when `body`
/// throws or returns a non-Void value.
@inline(__always)
func withOptionalCString<R>(_ s: String?, _ body: (UnsafePointer<CChar>?) throws -> R) rethrows -> R {
    guard let s else { return try body(nil) }
    return try s.withCString { try body($0) }
}
